### ISSUE TITLE
The `{type}` field in user actions log is untranslatable

### ISSUE THREAD

**Issue by nikosdion** [akeeba/engage#32]
## TL;DR – What we have to do

1. Create per-type language keys in the `actionlog` plugin, e.g. `COM_ENGAGE_USERLOG_COMMENT_CREATE_ARTICLE` instead of `COM_ENGAGE_USERLOG_COMMENT_CREATE`.
1. Modify each and every `actionlog` method to use the specific language key instead of the generic if such a language key exists. In this case the `type` argument must NOT be present in the array we send to the log. We can implement this with a helper method `prepareLanguageKey(string $genericKey, array &$info): string`
1. Upgrade existing UAL entries in the plugin's post-installation method.

## The surprisingly complex problem

The problem here is that we cannot return a translated noun from the content plugin. This works in English but not in Greek or other languages where you have genders and cases for nouns.

Think about the common pattern "submitted a comment on {type} such and such". In English you can easily interpolate a type to create grammatically correct phrases like "submitted a comment on article such and such" or "submitted a comment on contact such and such". In Greek the common pattern would be "υποβλήθηκε ένα σχόλιο για το {type} τάδε" which can become grammatically incorrect. For example "υποβλήθηκε ένα σχόλιο για το _περιεχόμενο_ τάδε" (correct gender) but "υποβλήθηκε ένα σχόλιο για το _επαφή_ τάδε" (incorrect gender on the article before the noun).

To further complicate things, Joomla stores the translation key and a JSON-encoded array of string interpolation parameters in the database. When the user actions are displayed we are not guaranteed to have any language file loaded except the one of the `actionlog` plugin. 

The problem with this approach is that whatever we put in the interpolation parameters array will be displayed _verbatim_ when the UAL is rendered. If we were to translate the type so that we could render gnarly but grammatically correct phrases like "submitted a comment on a content item of type “article“ which has the title “such and such“" we **still** have a problem! Think about the user displaying the page in Greek when filing the comment but the administrator being displayed in English. The end result would be the hilarious and absolutely wrong "submitted a comment on a content item of type “άρθρο“ which has the title “such and such“".

The only solution to bad grammar and bad language is... bad architecture :(

The `actionlog` plugin needs to have separate language strings for each and every content type, e.g. `COM_ENGAGE_USERLOG_COMMENT_CREATE_ARTICLE` instead of `COM_ENGAGE_USERLOG_COMMENT_CREATE`. If the specific language string does not exist we can still fall back to the generic `COM_ENGAGE_USERLOG_COMMENT_CREATE` knowing that the rendered result will be a language mix but what can you do...

The other problem with this UAL fix is that it won't retroactively change existing UAL entries but that's to be expected. We can, however, handle that in the post-installation script of the plugin – essentially "upgrade" existing entries to the new language string.

**Comment by nikosdion** [comment]
Upgrading the existing entries is not reliable without parsing every single UAL item that is likely ours. However, this is problematic due to the memory and time required. So we're simply not going to upgrade the old user log entries.
